Products/CMFPlone/relationhelper.py

```diff
@@ -148,101 +152,133 @@ def restore_relations(context=None, all_relations=None):
+        logger.info(f"Dropping {len(all_relations) - len(unique_relations)} duplicates")
         all_relations = unique_relations
 
     intids = getUtility(IIntIds)
     for index, item in enumerate(all_relations, start=1):
         if not index % 500:
-            logger.info(f'Restored {index} of {len(all_relations)} relations...')
+            logger.info(f"Restored {index} of {len(all_relations)} relations...")
 
+        # source object for UID exists.
         try:
-            source_obj = uuidToObject(item['from_uuid'])
+            source_obj = uuidToObject(item["from_uuid"])
         except KeyError:
             # brain exists but no object
             source_obj = None
+        finally:
+            if not source_obj:
+                logger.info(f'No source object found for UID {item["from_uuid"]}.')
+                continue
+
+        # target object for UID exists.
         try:
-            target_obj = uuidToObject(item['to_uuid'])
+            target_obj = uuidToObject(item["to_uuid"])
         except KeyError:
             # brain exists but no object
             target_obj = None
+        finally:
+            if not target_obj:
+                logger.info(f'No target object found for UID {item["to_uuid"]}.')
+                # The source_obj will be updated to remove the broken relation below.
 
-        if not source_obj:
-            logger.info(f'{item["from_uuid"]} is missing')
-            continue
-
-        if not target_obj:
-            logger.info(f'{item["to_uuid"]} is missing')
-            continue
-
+        # source_obj and target_obj are dexterity content types.
         if not IDexterityContent.providedBy(source_obj):
-            logger.info(f'{source_obj} is no dexterity content')
+            logger.info(f"{source_obj} is no dexterity content")
             continue
-
         if not IDexterityContent.providedBy(target_obj):
-            logger.info(f'{target_obj} is no dexterity content')
-            continue
+            logger.info(f"{target_obj} is no dexterity content")
 
-        from_attribute = item['from_attribute']
+        # Confirm intid for target_obj exists.
         try:
             to_id = intids.getId(target_obj)
         except KeyError as e:
-            logger.warning(f'No intid for {target_obj}')
-            continue
+            logger.warning(f"No intid for {target_obj}")
+            to_id = None
 
+        # Postpone linkintegrity check
+        from_attribute = item["from_attribute"]
         if from_attribute == referencedRelationship:
-            # Ignore linkintegrity for now. We'll rebuilt it at the end!
-            update_linkintegrity.add(item['from_uuid'])
+            update_linkintegrity.add(item["from_uuid"])
             continue
 
+        # Working copy relations
         if HAS_ITERATE and from_attribute == ITERATE_RELATION_NAME:
             # Iterate relations are not set as values of fields
             relation = StagingRelationValue(to_id)
```

Comment:
`to_id` could be `None` after your changes.  I think no staging relation should be created then.